libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    public class DialogContext
+    {
+        private object _finalResult;
+
+        public DialogSet Dialogs { get; set; }
+        public ITurnContext Context { get; set; }
+        public Stack<DialogInstance> Stack { get; set; }
+
+        /// <summary>
+        /// Creates a new DialogContext instance.
+        /// </summary>
+        /// <param name="dialogs">Parent dialog set.</param>
+        /// <param name="context">Context for the current turn of conversation with the user.</param>
+        /// <param name="stack">Current dialog stack.</param>
+        public DialogContext(DialogSet dialogs, ITurnContext context, Stack<DialogInstance> stack)
+        {
+            Dialogs = dialogs;
+            Context = context;
+            Stack = stack;
+        }
+
+        /// <summary>
+        /// Returns the cached instance of the active dialog on the top of the stack or `undefined` if the stack is empty.
+        /// </summary>
+        public DialogInstance Instance
+        {
+            get
+            {
+                if (Stack.Any())
+                {
+                    return Stack.Peek();
+                }
+                return null;
+            }
+        }
+
+        public DialogResult DialogResult
+        {
+            get
+            {
+                return new DialogResult
+                {
+                    Active = Stack.Count > 0,
+                    Result = _finalResult
+                };
+            }
+        }
+
+
+        /// <summary>
+        /// Pushes a new dialog onto the dialog stack.
+        /// </summary>
+        /// <param name="dialogId">ID of the dialog to start.</param>
+        /// <param name="dialogArgs">(Optional) additional argument(s) to pass to the dialog being started.</param>
+        public async Task Begin(string dialogId, object dialogArgs = null)
```

Comment:
If we have a Dialog STack, as defined by the code comments, why are the semantics "Begin"? 

A stack should have push / pop semantics.

Reply:
This library was primarily designed to be familiar to v3 Node developers and these are the names we used in v3.  I would probably use different names in some cases today but then it wouldn't be familiar and the names actually make sense once you understand how everything works.  I don't see any reason to change them.